### PR TITLE
DIV-5621 added solApplyForDA event

### PR DESCRIPTION
**Pull request by yarpo**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/DIV-5621

### Change description ###
Added solApplyForDA event

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: definitions/divorce/json/AuthorisationCaseEvent.json [hunk 1]
```diff
@@ -4627,12 +4627,12 @@
     "CRUD": "R"
   },
   {
-  "LiveFrom": "25/06/2019",
```
Comment (yarpo) [inline definitions/divorce/json/AuthorisationCaseEvent.json]: I haven't done it. Just result of execution `yarn generate-json`

File: definitions/divorce/json/CaseField.json [hunk 2]
```diff
@@ -3395,5 +3395,14 @@
...
     "FieldType": "TextArea",
     "SecurityClassification": "Public"
+  },
+  {
+    "LiveFrom": "23/09/2019",
+    "CaseTypeID": "DIVORCE",
+    "ID": "ApplyForDecreeAbsolute",
+    "Label": "Do you want to finalise your divorce?",
+    "FieldType": "YesOrNo",
+    "RegularExpression": "^Yes$",
```
Comment (yarpo) [inline definitions/divorce/json/CaseField.json]: Should I do it this way?

AC:
```
Validation is handled in the back end to ensure the event can not be completed if No is selected. In this case a message is displayed to the user saying "You have selected 'No'. You must select 'Yes' to apply for Decree Absolute"
```

or I shall I add callback url for aboutEvent? Or is there any other way?

File: definitions/divorce/json/CaseTypeTab.json [hunk 3]
```diff
@@ -1753,7 +1753,6 @@
     "TabLabel": "Payment",
     "TabDisplayOrder": 7,
     "CaseFieldID": "D8HelpWithFeesReferenceNumber",
-    "UserRole": "caseworker-divorce-courtadmin",
```
Comment (yarpo) [inline definitions/divorce/json/CaseTypeTab.json]: It was done automatically? Shall I revert it?

Comment (qzhou-hmcts) [reply definitions/divorce/json/CaseTypeTab.json]: Yes please revert all these changes, probably your base excel was not the latest one

File: definitions/divorce/json/FixedLists.json [hunk 4]
```diff
@@ -897,84 +897,72 @@
     "LiveFrom": "01/01/2019",
     "ID": "DnRefusalOptionEnum",
     "ListElementCode": "reject",
-    "ListElement": "Reject",
-    "DisplayOrder": 1
+    "ListElement": "Reject"
```
Comment (qzhou-hmcts) [inline definitions/divorce/json/FixedLists.json]: Revert these changes too

File: definitions/divorce/json/CaseEventToFields.json [hunk 5]
```diff
@@ -9109,5 +9109,18 @@
...
+  },
+  {
+    "LiveFrom": "23/09/2019",
+    "CaseTypeID": "DIVORCE",
+    "CaseEventID": "solApplyForDA",
+    "CaseFieldID": "ApplyForDecreeAbsolute",
+    "PageFieldDisplayOrder": 1,
+    "DisplayContext": "MANDATORY",
+    "PageID": "petsolapplyforda",
+    "PageLabel": "Do you want to finalise your divorce?",
```
Comment (qzhou-hmcts) [inline definitions/divorce/json/CaseEventToFields.json]: Content issue, this probably is not the solicitor divorce.

File: definitions/divorce/json/CaseEventToFields.json [hunk 6]
```diff
@@ -9109,5 +9109,18 @@
...
     "ShowSummaryChangeOption": "Yes"
+  },
+  {
+    "LiveFrom": "23/09/2019",
+    "CaseTypeID": "DIVORCE",
+    "CaseEventID": "solApplyForDA",
+    "CaseFieldID": "ApplyForDecreeAbsolute",
+    "PageFieldDisplayOrder": 1,
+    "DisplayContext": "MANDATORY",
+    "PageID": "petsolapplyforda",
```
Comment (alexwakeman) [inline definitions/divorce/json/CaseEventToFields.json]: Tend to use camelCase on PageID for readability

File: definitions/divorce/json/CaseField.json (hunk 2 above)
Comment (alexwakeman) [reply definitions/divorce/json/CaseField.json]: Would need to test it and see if this works. Hopefully we can just use this.

File: definitions/divorce/json/FixedLists.json [hunk 7]
```diff
@@ -897,84 +897,72 @@
...
   },
   {
     "LiveFrom": "01/01/2019",
     "ID": "DnRefusalOptionEnum",
     "ListElementCode": "moreInfo",
-    "ListElement": "Get more information / Evidence",
-    "DisplayOrder": 2
```
Comment (alexwakeman) [inline definitions/divorce/json/FixedLists.json]: Are we meant to drop the `DisplayOrder` entry?

File: definitions/divorce/json/CaseField.json (hunk 2 above)
Comment (alexwakeman) [reply definitions/divorce/json/CaseField.json]: Dynamic fields to use would be: `${DNApprovalDate}`, `${D8DerivedPetitionerCurrentFullName}` and `${D8DerivedRespondentCurrentName}

File: definitions/divorce/json/CaseField.json [hunk 8]
```diff
@@ -3395,5 +3395,14 @@
     "Label": "Additional information",
     "FieldType": "TextArea",
     "SecurityClassification": "Public"
+  },
+  {
+    "LiveFrom": "23/09/2019",
+    "CaseTypeID": "DIVORCE",
+    "ID": "ApplyForDecreeAbsolute",
+    "Label": "Once the court has checked that there are no outstanding applications or other matters which need to be resolved first, the decree nisi made on {dyinamic date} will be made absolute and the mariage between {dynamic data} and {dynamic data} will be ended",
```
Comment (alexwakeman) [inline definitions/divorce/json/CaseField.json]: Dynamic fields to use would be: `${DNApprovalDate}`, `${D8DerivedPetitionerCurrentFullName}` and `${D8DerivedRespondentCurrentName}`

**Comment by alexwakeman** [comment]
Not sure if we want to change the `ccd-definition-processor` git commit (sub-module version). You can fix it by checking which commit hash is used in master, the `cd` into your submodule and set the git HEAD to the same as master, re-commit and push.

File: definitions/divorce/json/CaseEventToFields.json (hunk 5 above)
Comment (gavinclark) [reply definitions/divorce/json/CaseEventToFields.json]: Can this change to "**Do you want to apply to finalise the divorce?**"

File: definitions/divorce/json/CaseField.json (hunk 2 above)
Comment (yarpo) [reply definitions/divorce/json/CaseField.json]: I have talked to Gavin. We are going to use checkbox instead.

File: definitions/divorce/json/CaseTypeTab.json (hunk 3 above)
Comment (yarpo) [reply definitions/divorce/json/CaseTypeTab.json]: reverted

File: definitions/divorce/json/CaseEvent.json [hunk 9]
```diff
@@ -2422,5 +2422,17 @@
     "PostConditionState": "*",
     "SecurityClassification": "PUBLIC",
     "ShowSummary": "Yes"
+  },
+  {
+    "LiveFrom": "23/09/2019",
+    "CaseTypeID": "DIVORCE",
+    "ID": "solApplyForDA",
+    "Name": "Solicitior applies for DA",
```
Comment (gavinclark) [inline definitions/divorce/json/CaseEvent.json]: Can you please change this to "Apply for Decree Absolute"? 
That is the name of the event visible to the solicitor in CCD.

File: definitions/divorce/json/CaseField.json [hunk 10]
```diff
@@ -3395,5 +3395,14 @@
     "Label": "Additional information",
     "FieldType": "TextArea",
     "SecurityClassification": "Public"
+  },
+  {
+    "LiveFrom": "23/09/2019",
+    "CaseTypeID": "DIVORCE",
+    "ID": "ApplyForDecreeAbsolute",
+    "Label": "Once the court has checked that there are no outstanding applications or other matters which need to be resolved first, the decree nisi made on ${DNApprovalDate} will be made absolute and the mariage between ${D8DerivedPetitionerCurrentFullName} and ${D8DerivedRespondentCurrentName} will be ended.",
+    "FieldType": "FixedList",
```
Comment (gavinclark) [inline definitions/divorce/json/CaseField.json]: I think the field type should be MultiSelectList

**Comment by hosainnet** [comment]
Can you please `yarn setup`?  I don't thing the changes in `ccd-definition-processor` are expected here

File: definitions/divorce/json/CaseEvent.json (hunk 9 above)
Comment (gavinclark) [inline definitions/divorce/json/CaseEvent.json]: Will also need authorisation for this event -
[PETSOLICITOR] - CRU
caseworker-divorce-courtadmin_beta - R
caseworker-divorce-courtadmin - R
caseworker-divorce-superuser - R
caseworker-divorce-courtadmin-la - R

File: definitions/divorce/json/FixedLists.json [hunk 11]
```diff
@@ -976,5 +976,11 @@
     "ListElementCode": "other",
     "ListElement": "Enter free Text / Make free text order",
     "DisplayOrder": 6
+  },
+  {
+    "LiveFrom": "23/09/2019",
+    "ID": "petSolApplyDaList",
+    "ListElementCode": "yes",
```
Comment (gavinclark) [inline definitions/divorce/json/FixedLists.json]: This should be 'Yes' (capitalised) to be consistent with the previously populated fields in the 'Yes' Or 'No' format

File: definitions/divorce/json/AuthorisationCaseEvent.json [hunk 12]
```diff
@@ -3345,6 +3401,20 @@
     "UserRole": "caseworker-divorce-courtadmin_beta",
     "CRUD": "R"
   },
+  {
+    "LiveFrom": "25/09/2019",
+    "CaseTypeID": "DIVORCE",
+    "CaseEventID": "SUAwaitingDA",
+    "UserRole": "[PETSOLICITOR]",
```
Comment (alexwakeman) [inline definitions/divorce/json/AuthorisationCaseEvent.json]: shouldn't this be caseworker-divorce-superuser only? Solicitors don't use this event.

File: definitions/divorce/json/AuthorisationCaseEvent.json [hunk 13]
```diff
@@ -1210,6 +1224,20 @@
     "UserRole": "caseworker-divorce-courtadmin_beta",
     "CRUD": "CRU"
   },
+  {
+    "LiveFrom": "25/09/2019",
+    "CaseTypeID": "DIVORCE",
+    "CaseEventID": "updateContactDetails",
+    "UserRole": "[PETSOLICITOR]",
```
Comment (alexwakeman) [inline definitions/divorce/json/AuthorisationCaseEvent.json]: Same here - don't need this
